modules/redis/scanner.go

```diff
@@ -0,0 +1,254 @@
+// Package redis provides a zgrab2 Module that probes for redis services.
+// The default port for redis is TCP 6379, and it is a cleartext protocol
+// defined at https://redis.io/topics/protocol.
+// Servers can be configured to require (cleartext) password authentication,
+// which is omitted from our probe by default (pass --password <your password>
+// to supply one).
+// Further, admins can rename commands, so even if authentication is not
+// required we may not get the expected output.
+// However, we should always get output in the expected format, which is fairly
+// distinct. The probe sends a sequence of commands and checks that the response
+// is well-formed redis data, which should be possible whatever the
+// configuration.
+package redis
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zmap/zgrab2"
+)
+
+// Flags contains redis-specific command-line flags.
+type Flags struct {
+	zgrab2.BaseFlags
+
+	// TODO: Take a JSON/YAML file with a list of custom commands to execute?
+	// TODO: Take a JSON/YAML file with mappings for command names?
+	AuthCommand string `long:"auth-command" default:"AUTH" description:"Override the command used to authenticate. Ignored if no password is set."`
+	Password    string `long:"password" description:"Set a password to use to authenticate to the server. WARNING: This is sent in the clear."`
+	DoInline    bool   `long:"inline" description:"Send commands using the inline syntax"`
+	Verbose     bool   `long:"verbose" description:"More verbose logging, include debug fields in the scan results"`
+}
+
+// Module implements the zgrab2.Module interface
+type Module struct {
+}
+
+// Scanner implements the zgrab2.Scanner interface
+type Scanner struct {
+	config *Flags
+}
+
+// Scan holds the state for the scan of an individual target
+type Scan struct {
```

Comment:
It looks like a lot of modules have a `Scan` struct that somehow gets converted into a `Result` struct. Is this something that we should consider at the framework level? Should these be public or private?

Reply:
I usually use a `Scan` if I need to keep track of state for one specific scan -- basically, if the scan gets too big to have everything in local variables in the `Scanner.Scan` method.
There would by definition be a 1:1 relationship between the `Scan` instances and results (just like there would be between `ScanTarget` and results), but I don't necessarily see it as a conversion.

Continuing the "encapsulated local variables" analogy, it would make sense for the `Scan` struct to be private. Anything outside of the module should be using the Scanner interface.